### PR TITLE
Allow ACF Option Page ID specification

### PR DESCRIPTION
**Pull request by DanArpoika**
### Original Issue
Needed to be able to access ACF Option Pages data via GraphQL. Unless I was doing something wrong (very possible), `allWordpressAcfOptions` was not accessible in the GraphQL browser IDE. 

With verbose set to true the `wp-json/acf/v3/options` (v2 also) endpoint would return a 404 on `gatsby develop`.

The `wp-json/acf/v3/options/{option-page-id}` endpoint where {option-page-id} is equal to the post_id set in the Wordpress function:

```
/* ACF Option Page */
acf_add_options_page( array(
	'page_title' => 'Option Page',
	....
	'post_id'    => 'option-page-id'
) );
```

would return the proper data. 

### Proposed Changes
The proposed changes in this PR would allow you to specify the IDs of the option pages you may have in Wordpress ACF. This requires the ACF Option Pages IDs to follow `abc_def` (underscores) naming and ACF to REST API plugin to be set to v3 as v2 returns null or 
```
{
'option-page-id': null
}
```

## Example gatsby-config.js 
**EDIT: Added acfRestVersion
```
module.exports = {
  plugins: [
    {
      resolve: 'gatsby-source-wordpress',
      options: {
        // The base url to your WP site.
        baseUrl: 'example.test',
        // WP.com sites set to true, WP.org set to false
        hostingWPCOM: false,
        // The protocol. This can be http or https.
        protocol: 'http',
        // Use 'Advanced Custom Fields' Wordpress plugin
        useACF: true,
        acfRestVersion: 3, // default is 3
        // Specify 'Advanced Custom Fields' option page IDs (if any)
        acfOptionPageIds: ['option_page_1_id', 'option_page_2_id'],
        auth: {},
        // Set to true to debug endpoints on 'gatsby build'
        verboseOutput: true
      }
    },
  ],
}

```

`gatsby develop` output with verbose set to true would show something like follows:
```
=== [ Fetching wordpress__acf_options ] === http://example.test/wp-json/acf/v3/options/option-page-1-id
⠠ source and transform nodes -> wordpress__acf_options fetched : 1

=== [ Fetching wordpress__acf_options ] === http://example.test/wp-json/acf/v3/options/option-page-2-id
⠠ source and transform nodes -> wordpress__acf_options fetched : 1
```

You can then query GraphQL...

```
query acfOptionPageQuery {
  wordpressAcfOptions(wordpress_id: {eq: "options_page_1_id"}) {
    id
    acf {
      company_overview_questions {
        question
      }
    }
  }
}

// output 
{
  "data": {
    "wordpressAcfOptions": {
      "id": "42c8ce94-7077-5632-9098-a23be3eb0352",
      "acf": {
        "company_overview_questions": [
          {
            "question": "What is Brett's last name?"
          }
        ]
      }
    }
  }
}
```

If an Option Page ID is specified that does not exist, the Wordpress REST endpoint will return 
```
{
acf: [ ]
}
```

and you will only be able to query `id` or `wordpress_id` on it.

Thoughts, comments, suggestions? Thanks! 🍰

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit c2eec2d2a5761dba3be4fd194f780d5ec2f27447

https://app.netlify.com/sites/using-postcss-sass/deploys/5b4fe73fb312742c9d4ae64a

**Comment by Khristophor** [comment]
@DanArpoika I ran into a similar issue when building a site with an ACF Options page. A (sort of hacky) workaround is to set the ACF to REST plugin to v2, but this might have unintended consequences on the rest of your ACF fields.

The root of the issue seems to be
https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L462
and the v3 version of the Options route not being added correctly.

**Comment by DanArpoika** [comment]
@Khristophor 
Was that recently?

I spun up a new instance of Wordpress with ACF and ACF to REST plugin. Neither v2 setting and
`/wp-json/acf/v2/options` or v3 setting and `/wp-json/acf/v3/options` returned any proper data. Only `/wp-json/acf/v3/{option_page_post_id}` did.

**Comment by Khristophor** [comment]
@DanArpoika we've been working on a site since ~March with those settings. ACF to REST is set to v2,  we have an Options page added with `acf_add_options_page();` in the theme, and we're able to get data from `/wp-json/acf/v2/options`: 
screen shot 2018-07-19 at 12 16 28 pm

```
=== [ Fetching wordpress__acf_options ] === https://[url]/wp-json/acf/v2/options
⠄ source and transform nodes -> wordpress__acf_options fetched : 1
Fetching the wordpress__acf_options took: 301.968ms
```

```graphQL
  query optionsQuery {
    allWordpressAcfOptions {
      edges {
        node {
          fallback {
            source_url
          }
        }
      }
    }
  }
```
```json
{
  "data": {
    "allWordpressAcfOptions": {
      "edges": [
        {
          "node": {
            "fallback": {
              "localFile": {
                "childImageSharp": {
                  "original": {
                    "src": "/static/6abee100a761fe8c99df4b80fb67b9d5-e9dc532077aa8d5978a6ac9ebdc01bb0.jpg"
                  }
                }
              }
            }
          }
        }
      ]
    }
  }
}
```

Maybe it's because it's an explicitly set post ID?

**Comment by DanArpoika** [comment]
@Khristophor To be completely honest, I forgot you could do just `acf_add_options_page();` without setting specifics. I might have to rethink this whole process. 

To confirm, yes just adding a `acf_add_options_page('Test Options');` with valid fields rendered the proper data at `/wp-json/acf/v2/options`..

**Comment by Khristophor** [comment]
So it looks like this would be a good solution for using v3 with a custom-named Options page, but ideally there would be support for both v2 and a default v3. According to [this issue](https://github.com/airesvsg/acf-to-rest-api/issues/184), the default behaviour is to set up the route as `/wp-json/acf/v3/options/options`, but presumably people will need the ability to identify custom Options page(s) urls as well.

**Comment by DanArpoika** [comment]
@Khristophor Looks like I have some work to do! Thanks for the insight.

**Comment by Khristophor** [comment]
@DanArpoika this honestly might be a better solution on the ACF to REST side, as it doesn't seem to expose it's Option page the same way as other endpoints. Looks like it would just need another route registered [here](https://github.com/airesvsg/acf-to-rest-api/blob/master/v3/lib/endpoints/class-acf-to-rest-api-options-controller.php#L15). [Gatsby uses the _links property](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-wordpress/src/fetch.js#L381) to determine a valid WordPress REST route, but the ACF to REST plugin doesn't expose them for Options pages.

If the change is made on the Gatsby side, we can see which version is being used based on the current logic (v2 vs v3), but I think we'll still need some custom setting for configured Options pages.

**Comment by DanArpoika** [comment]
@Khristophor I agree that it doesn't expose the option page the same way as it's other endpoints. However, I'm not sure how it could be done on the ACF to REST side since there is no way to retrieve all created options pages IDs. Either way would require setting configuration, but probably easier on the Gatsby side.

**Comment by DanArpoika** [comment]
@Khristophor I added support for both v2 and v3 of ACF to REST. Mind taking a look and letting me know your thoughts?

Also, not sure why the Travis CI build failed for something else unrelated.

**Comment by DanArpoika** [comment]
With the latest commit to this PR, you can now specify both the version of ACF to REST you are using, as well as any uniquely defined ACF Option Page's (OPs) in Wordpress.

Only V3 of ACF to REST will return data for uniquely defined ACF OPs and therefore we only add these routes if `acfRestVersion` is set to 3. 

The new code will combine the default option page data with any unique ACF OPs data.

You can query in GraphQL like below: 
  ```query optionsQuery {
    allWordpressAcfOptions {
      edges {
        node{
          company_overview_questions { ## From options page with post id option_page_1
            question
            type
            maturity_level
          }
          test_acf ## From default options page
        }
      }
    }
  }```

Query as you need!

A lot of our client builds we utilize multiple ACF OPs, and therefore this is a nice solution to being able to get that data and not have to use just one OP.

File: packages/gatsby-source-wordpress/src/normalize.js [hunk 1]
```diff
@@ -67,6 +67,26 @@ const normalizeACF = entities =>
 
 exports.normalizeACF = normalizeACF
 
+// Combine all ACF Option page data
+exports.combineACF = function(entities) {
+ let acfOptionData = {}
+ // Map each ACF Options object keys/data to single object
+ _.forEach(entities.filter((e) => e.__type === `wordpress__acf_options`), (e) => {
+   if(e[`acf`]) {
+     Object.keys(e[`acf`]).map((k) => acfOptionData[k] = e[`acf`][k])
```
Comment (pieh) [inline packages/gatsby-source-wordpress/src/normalize.js]: I'm not sure if I'm reading this correctly - but if we have 2 option pages both with same fieldnames, one would get overwritten? Perhaps we should have structure like this:
```
optionsNode
├─ optionsPageIdName1
|  └─ title (example field)
└─ optionsPageIdName2
   └─ title
```
(ignore this if I read this code wrong)

Comment (DanArpoika) [reply packages/gatsby-source-wordpress/src/normalize.js]: @pieh Thank you for the feedback.

I've pushed up a new commit that addresses both items in your review.

First, the ACF to REST version is now detected from the routes found. 

Secondly, I changed the structure to fit what you suggested. I neglected to think about duplicate field names. 

Now you would query like such:

```  query optionsQuery {
    allWordpressAcfOptions {
      edges {
        node{
          option_page_1 {
            test_acf
          }
          options {
            test_acf
          }
        }
      }
    }
  }
```

which would result in:
```
{
  "data": {
    "allWordpressAcfOptions": {
      "edges": [
        {
          "node": {
            "option_page_1": {
              "test_acf": "Test Text 2"
            },
            "options": {
              "test_acf": "Test Text 1"
            }
          }
        }
      ]
    }
  }
}
```

**Comment by jbsmith731** [comment]
@pieh any updates on getting this merged in?

**Comment by pieh** [comment]
I need some time to setup testing site. One thing that I suspect (but am not sure) is that we might not need to add `acfOptionPageIds: ['option_page_1_id', 'option_page_2_id'],` to `gatsby-config` - I feel like this could be extracted from `links` of main `options` endpoint?

**Comment by DanArpoika** [comment]
Hi @pieh, unfortunately neither v2/v3 of ACF to REST return links on the main `options` endpoint.

**Comment by pieh** [comment]
> Hi @pieh, unfortunately neither v2/v3 of ACF to REST return links on the main options endpoint.

Oh :(

In that case only thing needed is updating README to document that new config option.

**Comment by DanArpoika** [comment]
@pieh Should I create a new PR for that or update it in this one?

**Comment by pieh** [comment]
Please add that to this one

**Comment by Khristophor** [comment]
Thanks again @DanArpoika !

**Comment by pieh** [comment]
I'll also hijack this PR comments to link to https://github.com/gatsbyjs/gatsby/pull/6036 PR which has breaking change with how which routes are fetched by default - I would really appreciate Your input as users of wordpress plugins, because I feel like that change will serve well for majority of cases, but I might miss something there

**Pull request merged**